### ISSUE TITLE
java.lang.ExceptionInInitializerError on Samsung Devices with Android 5.x

### ISSUE THREAD

**Issue by rnevet** [ical4j/ical4j#125]
It seems that ical4j version 2.0.0 causes the following crash on Samsung devices with Android 5.x.x, it is related to the ServiceLoader usage. Ideas?

```
Fatal Exception: java.lang.ExceptionInInitializerError
       at net.fortuna.ical4j.model.Calendar.(SourceFile)
       at net.fortuna.ical4j.model.Calendar.(SourceFile)
       at net.fortuna.ical4j.data.CalendarBuilder$ContentHandlerImpl.startCalendar(SourceFile:307)
       at net.fortuna.ical4j.data.CalendarParserImpl.parse(SourceFile:119)
       at net.fortuna.ical4j.data.CalendarBuilder.build(SourceFile:198)
       at net.fortuna.ical4j.data.CalendarBuilder.build(SourceFile:178)
       at net.fortuna.ical4j.data.CalendarBuilder.build(SourceFile:165)
       ...
Caused by java.util.NoSuchElementException
       at java.util.ServiceLoader$ServiceIterator.next(ServiceLoader.java:210)
       at net.fortuna.ical4j.validate.AbstractCalendarValidatorFactory.(SourceFile)
       at net.fortuna.ical4j.model.Calendar.(SourceFile)
       at net.fortuna.ical4j.model.Calendar.(SourceFile)
       at net.fortuna.ical4j.data.CalendarBuilder$ContentHandlerImpl.startCalendar(SourceFile:307)
       at net.fortuna.ical4j.data.CalendarParserImpl.parse(SourceFile:119)
       at net.fortuna.ical4j.data.CalendarBuilder.build(SourceFile:198)
       at net.fortuna.ical4j.data.CalendarBuilder.build(SourceFile:178)
       at net.fortuna.ical4j.data.CalendarBuilder.build(SourceFile:165)
       ...
```

**Comment by benfortuna** [comment]
Have you tested in other Android platforms? It looks like it can't see the services config in the META-INF/ directory. 

Others have mentioned that some tools remove the META-INF directory when packaging as APK, I haven't yet tested myself though. I think the issue may still be open if you want to search for it.

**Comment by rnevet** [comment]
That does sound weird, actually this is crashing on a production app with many daily users the reports only indicate Samsung devices with Android 5.X. I did manage to reproduce it myself on an Galaxy S4 with Android 5.0.1, is there something additional I can test for?

**Comment by benfortuna** [comment]
I can't provide much more detail as I don't have an Android test environment right now. But if it does with some devices then it probably isn't a packaging issue.

**Comment by rnevet** [comment]
Could this be related to issue - https://github.com/ical4j/ical4j/issues/115 ?

**Comment by benfortuna** [comment]
#115 relates to use of OSGi, and specifically is caused by it's use of multiple class loaders that can't/load the resources from the META-INF directory. So whilst the symptom may be the same I don't think it would be the same cause, unless somehow Samsung phones (or maybe Android 5.x?) has locked down the classloader permissions or something like that.

**Comment by botty87** [comment]
Same problem of @rnevet! Only on Samsung device, S4 and Galaxy Tab 4, Android 5.X. Any other device has no problem.

**Comment by benfortuna** [comment]
The reason that 1.0.7 isn't affected is that it doesn't use service loader to loader the factories. It appears device loader is the problem so we'll be looking to replace it completely to resolve this issue.

**Comment by einatm** [comment]
Hey @benfortuna, if i'm using version 2.1.0 but do not wish to use the class loading since my app is having some difficulties loading and initializing the Calendar, is there a way to do that please?

**Comment by benfortuna** [comment]
Hi, I'm currently testing a replacement for service loader that doesn't rely on META-INF resources: #256

**Comment by einatm** [comment]
@benfortuna thank you very very much!

**Comment by rfc2822** [comment]
> no i didn't, it will mess things within my application.

I'm confused. Was that for me or for @benfortuna? It shouldn't mess your application, except you use the context class loader for anything special, which is very unlikely and you would know it :)

Seems that ServiceLoader just has been copied to the Android API because of Java compatibility, but they haven't adapted it to use the default class loader by default.

**Comment by benfortuna** [comment]
@rfc2822 is this context classloader workaround someone that should be added to the docs/wiki?

I'm aware of a few issues related to loading resources from the meta-inf path, so is this perhaps a solution for that?

**Comment by rfc2822** [comment]
> @rfc2822 is this context classloader workaround someone that should be added to the docs/wiki?

Yes, I'd add it. It makes ical4j perfectly usable on Android devices.

> I'm aware of a few issues related to loading resources from the meta-inf path, so is this perhaps a solution for that?

As far as I know – yes. ServiceLoader uses the context class loader to load the resources, but the context class loader is `null` on Android by default. You can simply set it to the current thread's class loader and everything will work.

I'm using this approach für ical4android and thus for DAVdroid and ICSdroid and don't have any problems with it.
